File: Backend/nuka-pos/src/main/java/com/nuka/nuka_pos/api/inventory/Inventory.java
package com.nuka.nuka_pos.api.inventory;

import com.nuka.nuka_pos.api.inventory.enums.InventoryStatus;
import com.nuka.nuka_pos.api.organization.Organization;
import com.nuka.nuka_pos.api.product.Product;
import com.nuka.nuka_pos.api.branch.Branch;
import jakarta.persistence.*;
import lombok.*;

import java.math.BigDecimal;
import java.time.LocalDateTime;

@Entity
@Getter
@Setter
@Builder
@Table(name = "inventory")
@NoArgsConstructor
@AllArgsConstructor
public class Inventory {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "organization_id", nullable = false)
    private Organization organization;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "product_id", nullable = false)
    private Product product;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "branch_id", nullable = false)
    private Branch branch;

    @Column(nullable = false)
    private Integer quantityAvailable;

    @Column(nullable = false)
    private Integer quantityReserved;

    @Column(nullable = false)
    private Integer quantitySold;

    @Column(precision = 10, scale = 2)
    private BigDecimal cost;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private InventoryStatus inventoryStatus;

    @Column(nullable = false, updatable = false)
    private LocalDateTime createdAt;

    @Column(nullable = false)
    private LocalDateTime updatedAt;

    @PrePersist
    public void prePersist() {
        this.createdAt = LocalDateTime.now();
        this.updatedAt = LocalDateTime.now();
    }

    @PreUpdate
    public void preUpdate() {
        this.updatedAt = LocalDateTime.now();
    }
}

File: Frontend/Nuka_POS/lib/models/inventory.dart
// lib/models/inventory.dart
class Inventory {
  final int id;
  final int organizationId;
  final int productId;
  final int branchId;
  final int quantityAvailable;
  final int quantityReserved;
  final int quantitySold;
  final double cost;
  final String inventoryStatus;
  final DateTime createdAt;
  final DateTime updatedAt;

  Inventory({
    required this.id,
    required this.organizationId,
    required this.productId,
    required this.branchId,
    required this.quantityAvailable,
    required this.quantityReserved,
    required this.quantitySold,
    required this.cost,
    required this.inventoryStatus,
    required this.createdAt,
    required this.updatedAt,
  });

  factory Inventory.fromJson(Map<String, dynamic> json) {
    return Inventory(
      id: json['id'] as int,
      organizationId: json['organizationId'] as int,
      productId: json['productId'] as int,
      branchId: json['branchId'] as int,
      quantityAvailable: json['quantityAvailable'] as int,
      quantityReserved: json['quantityReserved'] as int,
      quantitySold: json['quantitySold'] as int,
      cost: (json['cost'] as num).toDouble(),
      inventoryStatus: json['inventoryStatus'] as String,
      createdAt: DateTime.parse(json['createdAt'] as String),
      updatedAt: DateTime.parse(json['updatedAt'] as String),
    );
  }

  /// For POST/PUT payload
  Map<String, dynamic> toJsonForRequest() {
    return {
      'organization': {'id': organizationId},
      'product': {'id': productId},
      'branch': {'id': branchId},
      'quantityAvailable': quantityAvailable,
      'quantityReserved': quantityReserved,
      'quantitySold': quantitySold,
      'cost': cost,
      'inventoryStatus': inventoryStatus,
    };
  }
}
